draft-gpew-priv-ppm.md

```diff
@@ -672,247 +677,375 @@ independently.
+secret key indicated by `encrypted_input_share.config_id`. If not found, then it
+marks the report share as invalid with the error `hpke-unknown-config-id`.
+Otherwise, it decrypts the payload with the following procedure:
 
 ~~~
-context = SetupBaseR(helper_share.enc, sk, task_id ||
+context = SetupBaseR(encrypted_input_share.enc, sk, task_id ||
                      "ppm-00 input share" || 0x01 || server_role)
-input_share = context.Open(nonce || extensions, helper_share)
+
+input_share = context.Open(nonce || extensions,
+                           encrypted_input_share.payload)
 ~~~
 
-where `sk` is the HPKE secret key, `task_id` is `AggregateReq.task_id` and
-`server_role` is the role of the server (`0x02` for the leader and `0x03` for
-the helper; `0x01`, the client role, is used as the sender). `nonce` and
-`extensions` are obtained from the corresponding fields in `AggregateSubReq`. If
-decryption fails, then the sub-response consists of a "decryption error" alert.
-[See issue#57.] Otherwise, the helper handles the request for its plaintext
-input share `input_share` and updates its state as specified by the PPM
-protocol.
+where `sk` is the HPKE secret key, `task_id` is the task ID, `nonce` and
+`extensions` are the nonce and extensions of the report share respectively,
+and `server_role` is 0x02 if the aggregator is the leader and 0x03 otherwise.
+If decryption fails, the aggregator marks the report share as invalid with the
+error `hpke-decrypt-error`. Otherwise, it outputs the resulting `input_share`.
 
-After processing all of the sub-requests, the helper encrypts its updated state
-and constructs its response to the aggregate request.
+#### Input Share Validation {#input-share-batch-validation}
 
-#### Leader State
+Validating an input share will either succeed or fail. In the case of failure,
+the input share is marked as invalid with a ReportShareError of type `vdaf-prep-error`.
 
-The leader is required to buffer reports while waiting to aggregate them. The
-leader SHOULD NOT accept reports whose timestamps are too far in the future.
-Implementors MAY provide for some small leeway, usually no more than a few
-minutes, to account for clock skew.
+The validation checks are as follows.
 
-#### Helper State
+1. Check if the report has never been aggregated but is contained by
+   a batch that has been collected. If this check fails, the input share
+   is marked as invalid with the error `batch-collected`. This prevents
+   additional reports from being aggregated after its batch has already
+   been collected.
+2. Check if the report has already been aggregated. If this check fails,
+   the input share is marked as invalid with the error `report-replayed`.
+   This is the case if the report was used in a previous aggregate request
+   and is therefore a replay. An aggregator may also choose to mark an
+   input share as invalid with the  error `report-dropped` under the conditions
+   prescribed in {{anti-replay}}.
```

Comment:
Shouldn't the nonce check go here?

Reply:
Sorry, what nonce check?

Reply:
The check that the nonces aren't internally duplicated.